src/credentials.cpp

```diff
@@ -0,0 +1,315 @@
+		// Credentials file is a binary blob, so use streambuf iterator
+		std::string data((std::istreambuf_iterator<char>(*stream)), (std::istreambuf_iterator<char>()));
+		data = decrypt(data, build_key("global", get_system_username()));
+		if(data.empty() || data[0] != CREDENTIAL_SEPARATOR) {
+			ERR_CFG << "Invalid data in credentials file\n";
+			std::fill(data.begin(), data.end(), '\0');
+			return;
+		}
+		for(const std::string elem : utils::split(data, CREDENTIAL_SEPARATOR, utils::REMOVE_EMPTY)) {
+			size_t at = elem.find_last_of('@');
+			size_t eq = elem.find_first_of('=', at + 1);
+			if(at != std::string::npos && eq != std::string::npos) {
+				credentials.emplace_back(elem.substr(0, at), elem.substr(at + 1, eq - at - 1), unescape(elem.substr(eq + 1)));
+			}
+		}
+		std::fill(data.begin(), data.end(), '\0');
+	}
+
+	void save_credentials()
+	{
+		if(!remember_password()) {
+			filesystem::delete_file(filesystem::get_credentials_file());
+			return;
+		}
+		std::ostringstream credentials_data;
+		for(const auto& cred : credentials) {
+			credentials_data.put(CREDENTIAL_SEPARATOR);
+			std::copy(cred.username.begin(), cred.username.end(), std::ostreambuf_iterator<char>(credentials_data));
+			credentials_data.put('@');
+			std::copy(cred.server.begin(), cred.server.end(), std::ostreambuf_iterator<char>(credentials_data));
+			credentials_data.put('=');
+			std::string key_escaped = escape(cred.key);
+			std::copy(key_escaped.begin(), key_escaped.end(), std::ostreambuf_iterator<char>(credentials_data));
+		}
+		credentials_data.put(CREDENTIAL_SEPARATOR);
+		try {
+			filesystem::scoped_ostream credentials_file = filesystem::ostream_file(filesystem::get_credentials_file());
+			std::string encrypted = encrypt(credentials_data.str(), build_key("global", get_system_username()));
+			credentials_file->write(encrypted.c_str(), encrypted.size());
+		} catch(filesystem::io_exception&) {
+			ERR_CFG << "error writing to credentials file '" << filesystem::get_credentials_file() << "'" << std::endl;
+		}
+		size_t n = credentials_data.tellp();
+		credentials_data.seekp(0, std::ios::beg);
+		std::fill_n(std::ostreambuf_iterator<char>(credentials_data), n, '\0');
+	}
+}
+
+// TODO: Key-stretching (bcrypt was recommended)
+std::string build_key(const std::string& server, const std::string& login)
+{
+	std::ostringstream out;
+	out << login << get_system_username() << server;
+	while(out.tellp() < 32) {
+		out.put('x' ^ out.tellp());
+	}
+	return out.str();
+}
+
+static std::vector<unsigned char> rc4_crypt(const std::vector<unsigned char>& text, const std::vector<unsigned char>& key)
+{
+	RC4_KEY cipher_key;
+	RC4_set_key(&cipher_key, key.size(), key.data());
+	std::vector<unsigned char> result;
+	size_t blocks = text.size() / key.size();
+	size_t extra = text.size() % key.size();
+	for(size_t i = 0; i < blocks * key.size(); i += key.size()) {
+		result.resize(result.size() + key.size(), '\0');
```

Comment:
Please create the `result` vector with the right size from the start instead of resizing it.

In line 243 you can calculate the offset with

    size_t i = blocks * block_size;